Decouple logs output color from the logging level

healthy-pod opened the pull request:
Previously, each logging level was coupled with a specific
output color. This isn't always ideal because we might need
to log at a certain level but use a custom color (based on
user feedback / see linked issue).

This patch decouples the logs output color from the logging
level.

Fixes #10519

<!---
Thank you for your soon to be pull request. Before you submit this, please
double check to make sure that you've added a news file fragment. In pip we
generate our NEWS.rst from multiple news fragment files, and all pull requests
require either a news file fragment or a marker to indicate they don't require
one.

To read more about adding a news file fragment for your PR, please check out
our documentation at: https://pip.pypa.io/en/latest/development/contributing/#news-entries
-->

pradyunsg commented:
Hurrah! Work for #10421 is making these little things easier as well!

healthy-pod commented:
Fixed (at least I think I did) a pre-commit issue in the last change. It's my first contribution to `pip` so feel free to let me know if there's anything I need to update!

DiddiLeija commented:
The linters are going well. The tests are currently failing because of #10742.

pradyunsg commented:
Doing the close-reopen dance, in case GH actions runs on a simulated merge commit.

webknjaz commented on src/pip/_internal/resolution/resolvelib/factory.py:
```diff
@@ -692,7 +692,7 @@ def describe_trigger(parent: Candidate) -> str:
         )
 
-        logger.info(msg)
+        logger.critical(msg, extra={"color": "black"})
```
Why should this be black? A lot of terminals have black/dark background. Having black on black wouldn't be consumable, no?

webknjaz replied:
Looking into this whole branchy structure, why not set a color variable dynamically and initialize `Style` once with that var?

pradyunsg replied:
> Why should this be black?

It corresponds to ANSI 0: https://rich.readthedocs.io/en/stable/appendix/colors.html#appendix-colors

pradyunsg replied:
> Looking into this whole branchy structure, why not set a color variable dynamically and initialize Style once with that var?

That can be done separately in a follow up. I'd like to not increase the scope of this PR, especially given that this is a first time contributor! :)

healthy-pod replied:
> Why should this be black? A lot of terminals have black/dark background. Having black on black wouldn't be consumable, no?

Right, I thought the terminal emulator would deal with that if the background isn't white? Do we have a way to check the current terminal output color?

webknjaz replied:
Dunno, that's why I pointed it out. @pradyunsg do I understand correctly, that this is the color of the text, not the background?

pradyunsg replied:
Yup. This will always be the default output color of the terminal.

webknjaz commented on src/pip/_internal/utils/logging.py:
```diff
@@ -162,7 +162,11 @@ def emit(self, record: logging.LogRecord) -> None:
+            # If a custom color is passed use it, otherwise use default colors.
+            color_attribute = "color"
+            if hasattr(record, color_attribute):
+                style = Style(color=getattr(record, color_attribute))
```
It's usually better to follow EAFP in Python. As in, access the attribute with a dot-syntax and not a dynamic function call. Could go for try/except in this case.

pradyunsg replied:
I find that this is clearer.

webknjaz replied:
Wouldn't it confuse things like mypy?

healthy-pod replied:
Actually mypy will be confused by dot-syntax because `record` class doesn't have a `color` attribute by default so mypy screams if you try to access `color` using `record.color`.
FWIW, this is how `color` or any other extra attribute are assigned to the record:
```
        if extra is not None:
            for key in extra:
                if (key in ["message", "asctime"]) or (key in rv.__dict__):
                    raise KeyError("Attempt to overwrite %r in LogRecord" % key)
                rv.__dict__[key] = extra[key]
```
https://github.com/python/cpython/blob/3.10/Lib/logging/__init__.py#L1593:L1597

webknjaz reviewed (commented):
@healthy-pod why did you add a "bugfix" change note? This seems to be a feature. Also, is the note text empty by accident?

On a side note: it may be worth splitting the addition of the mechanism that implements attaching colors to messages and using it. I'd imagine that the first part could benefit from some sort of automatic tests.

P.S. I see you're allowing color strings — would allowing to set the whole style too be useful?

pradyunsg commented:
Hiya @healthy-pod! Could you rebase this PR? If you're not familiar with the idea, you might find this article to be a [useful read](https://hackernoon.com/beginners-guide-to-interactive-rebasing-346a3f9c3a6d). :)

healthy-pod commented:
> Hiya @healthy-pod! Could you rebase[1](#user-content-fn-1-7d34077c9eb879f296994e45fb299e8c) this PR?
> 
> This branch was created at a point where the CI on `main` was broken, and thus, can't be merged as-is.


Done

pradyunsg commented:
I swear our CI is usually not a fickle beast like this. Ok, maybe it is...

Anyway, @healthy-pod could you rebase this _again_?

healthy-pod commented:
>  Anyway, @healthy-pod could you rebase this again?

Done

pradyunsg commented:
Oh hey, look! The test suite works, and there's tests that caught that you're spewing to stderr instead of stdout now. You'll likely need to fix those! :)

pradyunsg commented:
Hiya @healthy-pod! A gentle nudge on this -- this PR needs a minro update. If you won't have time to address this soon, please let me know and I'll pick it up. :)

healthy-pod commented:
> Hiya @healthy-pod! A gentle nudge on this -- this PR needs a minro update. If you won't have time to address this soon, please let me know and I'll pick it up. :)

Thanks for the reminder @pradyunsg! I assume "critical" goes to stderr while "info" goes to stdout. I changed `logger.info` to `logger.critical` in [factory.py](https://github.com/pypa/pip/pull/10763/files#diff-64990ce29c78c39f1164f70593aba9c8f39bdf7cf499e2749b09a6da5ce2901f) which caused some tests expecting some output at stdout to get it at stderr. Is my analysis correct? If so, do we want to change the tests in this case (ie change what they expect)? I wonder what the tests did when the line I changed was originally `logger.critical` (ie before it was even changed to `logger.info`).
